### ISSUE TITLE
Provide a method for unwrapping the underlying cause of a PropagatedCacheException

### ISSUE THREAD

**Issue by tburny** [headissue/cache2k#14]
When using read through access and an exception in the cache source occurs, a `PropagatedCacheException`(PCE) is thrown, which wraps the original exception in one or even multiple layers (for example if Caches are layered, Cache2: Cache1.get() -> Cache1.get()=PCE(UnderlyingException) => Cache2.get() = PCE(PCE(UnderlyingException))).

As a result PCE.getCause() cannot be used reliably to get the underlying original Exception.
Just skipping intermediate PCE's in PCE.getCause is not a good idea, as it would hide parts of the execution path in the stack trace.

Thus, PCE should have a method getOriginalCause(), like sketched here:
```java
public Throwable getOriginalCause() {
    while (t.getCause() instanceof PropagatedCacheException) {
        t = (Exception) t.getCause();
      }
      if (t.getCause() != null) {
        t = (Exception) t.getCause();
      }
      // Maybe throw (and chuck?) instead
      return t;

}
```
Moreover it would be much more straightforward if a cache could be configured to always throw the original cause by default.
The cache is nothing more than a transparent layer. Imagine a method `foo()` which may throw a `NullPointerException`:
```java
public void foo(Object o) {
Objects.reqireNonNull(o); // throws NPE if o == null
return o;
}
```
we can write a try/catch-block to catch it:

```java
try {
   foo();
} catch (NullPointerException e) {
// handle exception
}
```
Now if using `foo()` as cache source, by reading the source code, e.g. 

```java
CacheBuilder ... .source(this::foo)
```

or (expanded)
```java
CacheBuilder ... .source((e) -> {
Objects.requireNonNull(e);
return e;
})
```
one would expect a `NullPointerException` might be thrown. However, a PCE **wrapping** the NPE is thrown, which might be very unexpected due to the idea of transparency with read-through operation mode.
This also reduces complexity and improves readability:
```java
Cache<Object, Object> cache =
CacheBuilder.newInstance(Object.class, Object.class).source(this::foo).build();

try {
   cache.get(new Object());// with cache.source = foo()
} catch (NullPointerException e) {
// do something
}
```
is much more readable than

```java
try {
   cache.get(new Object());// with cache.source(this::foo)
} catch (NullPointerException|PropagatedCacheException ex) {
  if (ex instanceof PropagatedCacheException 
     // There might be a very bad exception not coming from the cache source -> no cause
    && ex.getCause() != null) {
    ex = ex.getCause()
  }
  if (!(ex instanceof NullPointerExcetion)) {
    // rethrow, something bad happened
    throw e;
  }
  // handle NPE
}
``` 

The programmer might even not have the possibility of a PCE in mind if using a cache.

**Comment by cruftex** [comment]
Dealing with exceptions is one of the more esoteric things with caching. I doubt that there will be the perfect solution.

About getOrginialCause():

With a sound layered program design each application should provide a proper interface to the layer using it.  If there is a need to react on exceptions "down below" two caches, its job of the layer using cache 2 to handle the propagatedcacheexception and throw something meaningful.

There may be use cases, but I don't think it is a good idea to put it into a general API. If there is need to do some magical stuff with exceptions, this can be done anyways.

About rethrowing the original exception:

In the presence of cached exceptions that is not a good idea, since the stack traces may be totally different. Even if the exception is not cached and a get() and getAll() access the same resource at the same time, would yield the identical exception.

So, throwing an exception with a wrong and confusing stack trace, is not a good idea either. Maybe construct a synthetic stacktrace by combining the current and the past one?! Hmm, this gets quite nasty....

The second thing to think about is, that it is important to know, that the exception was cached and did not just happen at the same time.

In summary:

I think there is no general perfect solution and the current one is the least troublesome. A good idea, which actually solves both of your mentioned problems, is to add the ability to add a custom "exception propagator".

**Comment by cruftex** [comment]
Just added https://github.com/headissue/cache2k/issues/15 for following this up.

**Comment by cruftex** [comment]
BTW:
Available in master / unreleased: An ExceptionPropagator can be set via CacheBuilder.exceptionPropagater()

Comments pls. in issue #15
